src/gtc/numpy/npir.py

```diff
@@ -22,147 +22,164 @@
 from gtc import common
 
 
-@eve.utils.noninstantiable
-class Expr(common.Expr):
-    pass
+HorizontalExtent = Tuple[Tuple[int, int], Tuple[int, int]]
 
 
-class Literal(common.Literal, Expr):
-    @validator("dtype")
-    def is_defined(cls, dtype: common.DataType) -> common.DataType:
-        undefined = [common.DataType.AUTO, common.DataType.DEFAULT, common.DataType.INVALID]
-        if dtype in undefined:
-            raise ValueError("npir.Literal may not have undefined data type.")
-        return dtype
+# --- Misc ---
+class AxisName(eve.StrEnum):
+    I = "I"  # noqa: E741 (ambiguous variable name)
+    J = "J"
+    K = "K"
+
+
+# --- Decls ---
+@eve.utils.noninstantiable
+class Decl(eve.Node):
+    name: eve.SymbolName
+    dtype: common.DataType
 
 
-class Cast(common.Cast[Expr], Expr):
+class ScalarDecl(Decl):
+    # Scalar per grid point
+    # Locals never have data_dims
+    # Used for API scalar parameters and oir local scalars
     pass
 
 
-class NumericalOffset(eve.Node):
-    value: int
+class FieldDecl(Decl):
+    # General field shared across HorizontalBlocks
+    dimensions: Tuple[bool, bool, bool]
+    data_dims: Tuple[int, ...] = eve.field(default_factory=tuple)
+    extent: HorizontalExtent
 
 
-class AxisName(eve.StrEnum):
-    I = "I"  # noqa: E741 (ambiguous variable name)
-    J = "J"
-    K = "K"
+class TemporaryDecl(Decl):
+    # Temporary field shared across HorizontalBlocks
+    data_dims: Tuple[int, ...] = eve.field(default_factory=tuple)
+    offset: Tuple[int, int]
+    boundary: Tuple[int, int]
```

Comment:
What's the actual meaning of the numbers here?

Reply:
Offset is the origin of the temporary field, and boundary (changed to padding) is the extra padding needed compared to the compute domain.